File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private HashMap<Integer,Integer> mapp = new HashMap<>();
    private int preIndex = 0;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for(int i=0;i<inorder.length;i++){
            mapp.put(inorder[i],i);
        }

        return build(preorder,0,preorder.length-1);
    }

    private TreeNode build(int[] preorder,int left,int right){
        if(left > right){
            return null;
        }

        int val = preorder[preIndex++];
        TreeNode tempNode = new TreeNode(val);
        int mid = mapp.get(val);

        tempNode.left = build(preorder,left,mid-1);
        tempNode.right = build(preorder,mid+1,right);
        
        return tempNode;
    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        ans = []
        self.preorder(root,ans)
        return ",".join(ans)
    
    def preorder(self,root,ans):
        if root is None:
            ans.append("N")
            return
        ans.append(str(root.val))
        self.preorder(root.left,ans)
        self.preorder(root.right,ans)

    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        arr= data.split(",")
        self.index = 0 
        return self.build(arr)

    def build(self,arr):
        if arr[self.index] == "N":
            self.index += 1
            return None
        
        root = TreeNode(int(arr[self.index]))
        self.index += 1
        root.left = self.build(arr)
        root.right = self.build(arr)

        return root


        